src/runtime.c

```diff
@@ -536,7 +528,15 @@ int main(int argc, char *argv[]) {
      * functionality specifically for builds used by appimaged.
      */
     if (getenv("TARGET_APPIMAGE") == NULL) {
-        strcpy(appimage_path, "/proc/self/exe");
+        // for some reason, `fopen("/proc/self/exe", "rb")` tries to open
+        // the dynamic linker when running under gcompat, even though `readlink()`
+        // gives the right result, so use `readlink()` here
+        ssize_t len = readlink("/proc/self/exe", appimage_path, sizeof(appimage_path));
+        if (len < 0) {
```

Comment:
You also have to check whether the buffer is too small by comparing `len` to `sizeof(appimage_path)` (see `man 2 readlink`).

Reply:
Isn't that redundant?  If the buffer's too small, then the path size exceeded the buffer, but the buffer is `PATH_MAX` bytes large, and the symlink in question is `/proc/self/exe`, which I think should always be < PATH_MAX (excluding the null terminator).

Reply:
`PATH_MAX` is just some random size macro (typically 4096) in the POSIX API. The name is really misleading. Filesystems do not actually limit paths to that size, they have their own specific limits. It may not be too realistic but it is definitely possible that the path would be truncated because the buffer is too small. Therefore, a check seems necessary.

That's by the way why I prefer to use `realpath`, which just `malloc`s a buffer for you.

C is really annoying when working with actual data... I strongly prefer C++ nowadays.